tfx_bsl/beam/run_inference.py

```diff
@@ -271,19 +344,21 @@ def update(self, elements: List[Union[tf.train.Example,
       self._inference_request_batch_byte_size.update(
           sum(element.ByteSize() for element in elements))
 
-  def __init__(self, inference_spec_type: model_spec_pb2.InferenceSpecType):
+  def __init__(
+      self,
+      fixed_inference_spec_type: model_spec_pb2.InferenceSpecType = None
+  ):
     super(_BaseDoFn, self).__init__()
     self._clock = None
-    self._metrics_collector = self._MetricsCollector(inference_spec_type)
+    self._metrics_collector = self._MetricsCollector(fixed_inference_spec_type)
```

Comment:
Can this be moved to setup_model() ? So that we can know the proximity and operation_type for model streaming case.

Reply:
My concern with this is that we might run into issues initializing Beam metrics outside of `__init__`. That might be a question for someone with more experience with Beam/Dataflow. 

For the purposes of metrics collection, I think the end goal here is to collect metrics per operation type (e.g. classify, regress, ...). The `fixed_inference_spec_type` is only used here to determine the operation type so we could get rid of that and use a different solution where we initialize unique metric collectors for each operation type (which does not require knowing the inference spec). Then at runtime we choose which metric collector to use based on the inference spec (which will be available).

e.g. here we would have something like:
```py
self._classify_metrics = self._MetricsCollector(OperationType.CLASSIFICATION, _METRICS_DESCRIPTOR_IN_PROCESS)
self._regress_metrics = self._MetricsCollector(OperationType.REGRESSION, _METRICS_DESCRIPTOR_IN_PROCESS)
...
```

and maybe expose a new method:

```py
class _BaseDoFn(beam.DoFn):
  ...
  def _metrics_collector_for_inference_spec(inference_spec_type: InferenceSpecType) -> _MetricsCollector:
    ...
```

internally, we could replace:

```py
self._metrics_collector.update(...)
```

with

```py
self._metrics_collector_for_inference_spec(inference_spec).update(...)
```

Does this sound reasonable?

Reply:
It sounds good. Also note that we need to toggle on _METRICS_DESCRIPTOR_IN_PROCESS and _METRICS_DESCRIPTOR_CLOUD_AI_PREDICTION

Reply:
I found a slightly cleaner solution here: I refactored `MetricsCollector` to accept an operation type and proximity directly: https://github.com/hgarrereyn/tfx-bsl/blob/core/tfx_bsl/beam/run_inference.py#L299-L309

Each subclass DoFn is responsible for configuring the correct operation type and proximity (for this we don't need the inference spec).

I refactored the `MetricCollector` methods for readability so now there is:
- `update_inference` for updating inference metrics
- `update_model_load` + `commit_cached_metrics` for updating model loading metrics. The first function caches the metrics and the second function commits the metrics. In the fixed model case, `update_model_load` is called in `DoFn.setup` and `commit_cached_metrics` is called in `DoFn.finish_bundle` and I added some documentation explaining why. In the dynamic model case, both functions are called sequentially in `DoFn.process`.